Frameworks/Foundation/NSUndoManager.mm

```diff
@@ -14,18 +14,84 @@
 //
 //******************************************************************************
 
+#import <Starboard.h>
 #import <StubReturn.h>
 #import <Foundation/NSUndoManager.h>
+#import "_NSUndoManagerInternal.h"
 
 NSString* const NSUndoManagerGroupIsDiscardableKey = @"NSUndoManagerGroupIsDiscardableKey";
 
-@implementation NSUndoManager
+@implementation NSUndoManager {
+    StrongId<_NSUndoManagerStack> _undoStack;
+    StrongId<_NSUndoManagerStack> _redoStack;
+    NSUInteger _undoGroupingLevel;
+    NSUInteger _redoGroupingLevel;
+    NSUInteger _disableCount;
+    BOOL _isAutomaticGroupOpen;
+    BOOL _hasNestedGroup;
+}
+
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
+*/
+- (id)init {
+    if (self = [super init]) {
+        _runLoopModes = @[ NSDefaultRunLoopMode ];
+        _groupsByEvent = YES;
+
+        _disableCount = 0;
+        _undoGroupingLevel = 0;
+        _redoGroupingLevel = 0;
+
+        _undoStack.attach([[_NSUndoManagerStack alloc] init]);
+        _redoStack.attach([[_NSUndoManagerStack alloc] init]);
+    }
+    return self;
+}
+
+/**
+ @Status Interoperable
 */
 - (void)registerUndoWithTarget:(id)target selector:(SEL)aSelector object:(id)anObject {
-    UNIMPLEMENTED();
+    if (_disableCount == 0) {
+        _NSUndoBasicAction* undoCall =
+            [[_NSUndoBasicAction alloc] _initWithTarget:target selector:aSelector object:anObject modes:self.runLoopModes];
+        [self _registerUndoAction:undoCall];
+    }
+}
+
+- (void)_createAutomaticUndoGroup {
+    if (!_isAutomaticGroupOpen && _undoGroupingLevel == 0 && _groupsByEvent) {
+        _isAutomaticGroupOpen = YES;
+        [self beginUndoGrouping];
+    }
+}
+
+// Internal function for proper registration of the deferred call.
+- (void)_registerUndoAction:(_NSUndoBasicAction*)undoCall {
```

Comment:
nit: This should probably be above its call (does it generate a warning because this symbol hasn't been declared yet?). This ordering also, to me at least, feels more immediately intuitive